checks/http.go

```diff
@@ -0,0 +1,242 @@
+package checks
+
+import (
+	"encoding/json"
+	"log"
+	"strconv"
+	"time"
+	"net"
+	"net/http"
+	"io"
+	"io/ioutil"
+	"fmt"
+	"strings"
+	"bufio"
+	"bytes"
+	"regexp"
+	"compress/gzip"
+)
+
+// HTTPResult struct
+type HTTPResult struct {
+	DNS        *float64 `json:"dns"`
+	Connect    *float64 `json:"connect"`
+	Send       *float64 `json:"send"`
+	Wait       *float64 `json:"wait"`
+	Recv       *float64 `json:"recv"`
+	Total      *float64 `json:"total"`
+	DataLength *float64 `json:"dataLength"`
+	StatusCode *float64 `json:"statusCode"`
+	Error      *string  `json:"error"`
+}
+
+// RaintankProbeHTTP struct.
+type RaintankProbeHTTP struct {
+	Host        string      `json:"host"`
+	Path        string      `json:"path"`
+	Port        string      `json:"port"`
+	Method      string      `json:"method"`
+	Headers     string      `json:"headers"`
+	ExpectRegex string      `json:"expectRegex"`
+	Result      *HTTPResult `json:"-"`
+}
+
+// NewRaintankHTTPProbe json check
+func NewRaintankHTTPProbe(body []byte) (*RaintankProbeHTTP, error) {
+	p := RaintankProbeHTTP{}
+	err := json.Unmarshal(body, &p)
+	if err != nil {
+		log.Fatalf("failed to parse settings. %v", err.Error())
+		return nil, err
+	}
+	if port, err := strconv.ParseInt(p.Port, 10, 32); err != nil || port < 1 || port > 65535 {
+		log.Fatal("failed to parse settings. Invalid port")
+		return nil, err
+	}
+	if !strings.HasPrefix(p.Path, "/") {
+		return nil, fmt.Errorf("invalid path. must start with /")
+	}
+	return &p, nil
+}
+
+// Results interface
+func (p *RaintankProbeHTTP) Results() interface{} {
+	return p.Result
+}
+
+
+// Run checking
+func (p *RaintankProbeHTTP) Run() error {
+	p.Result = &HTTPResult{}
+	
+	if p.Method == "" {
+		p.Method = "GET"
+	}
+	
+	// reader
+	url := fmt.Sprintf("http://%s:%s%s", p.Host, p.Port, p.Path)
+	request, err := http.NewRequest(p.Method, url, nil)
+	request.Header.Set("Connection", "close")
+	
+	// Parsing header (use fake request)
+	if p.Headers != "" {
+		headReader := bufio.NewReader(strings.NewReader("GET / HTTP/1.1\r\n" + p.Headers + "\r\n\r\n"))
+		dummyRequest, err := http.ReadRequest(headReader)
+		if err != nil {
+			msg := err.Error()//"failed to parse header."
+			p.Result.Error = &msg
+			return nil
+		}
+		
+		for key := range dummyRequest.Header {
+			request.Header.Set(key, dummyRequest.Header.Get(key))
+		}
+	}
+	
+	if err != nil {
+		msg := "connection closed"
+		p.Result.Error = &msg
+		return nil
+	}
+	
+	// Dialing
+	start := time.Now()
+	conn, err := net.Dial("tcp", p.Host+":"+p.Port)
+	if err != nil {
+		msg := err.Error()
+		p.Result.Error = &msg
+		return nil
+	}
+	duration := time.Since(start).Seconds() * 1000
+	p.Result.Connect = &duration
+	
+	// DNS
+	step := time.Now()
```

Comment:
The DNS needs to be queried before trying to connect.  When dialing the connection you need to use the IP address returned from the DNS request, so that the Dial process does not need to also perform a DNS lookup.